File: src/FSharpVSPowerTools/ResolveUnopenedNamespaceSmartTaggerProvider.cs
using System;
using System.ComponentModel.Composition;
using Microsoft.VisualStudio.Text;
using Microsoft.VisualStudio.Text.Editor;
using Microsoft.VisualStudio.Text.Operations;
using Microsoft.VisualStudio.Text.Tagging;
using Microsoft.VisualStudio.Utilities;
using Microsoft.VisualStudio.Shell;
using FSharpVSPowerTools.Refactoring;
using FSharpVSPowerTools.ProjectSystem;
using System.Diagnostics;
using Microsoft.VisualStudio.Shell.Interop;

namespace FSharpVSPowerTools
{
    [Export(typeof(IViewTaggerProvider))]
    [ContentType("F#")]
    [TagType(typeof(ResolveUnopenedNamespaceSmartTag))]
    [TextViewRole(PredefinedTextViewRoles.Editable)]
    public class ResolveUnopenedNamespaceSmartTaggerProvider : IViewTaggerProvider
    {
        [Import]
        internal VSLanguageService fsharpVsLanguageService = null;

        [Import]
        internal ITextDocumentFactoryService textDocumentFactoryService = null;

        [Import(typeof(SVsServiceProvider))]
        internal IServiceProvider serviceProvider = null;

        [Import]
        internal ITextUndoHistoryRegistry undoHistoryRegistry = null;

        [Import]
        internal ProjectFactory projectFactory = null;

        public ITagger<T> CreateTagger<T>(ITextView textView, ITextBuffer buffer) where T : ITag
        {
            // Only provide the smart tagger on the top-level buffer
            if (textView.TextBuffer != buffer) return null;

            var generalOptions = Setting.getGeneralOptions(serviceProvider);
            if (generalOptions == null || !generalOptions.ResolveUnopenedNamespacesEnabled) return null;

            var dte = serviceProvider.GetService(typeof(SDTE)) as EnvDTE.DTE;
            var vsVersion = VisualStudioVersionModule.fromDTEVersion(dte.Version);
            if (vsVersion == VisualStudioVersion.VS2015) return null;

            ITextDocument doc;
            if (textDocumentFactoryService.TryGetTextDocument(buffer, out doc))
            {
                var resolver = new UnopenedNamespaceResolver(doc, textView, undoHistoryRegistry.RegisterHistory(buffer),
                     fsharpVsLanguageService, serviceProvider, projectFactory);
                return new ResolveUnopenedNamespaceSmartTagger(buffer, serviceProvider, resolver) as ITagger<T>;
            }
            
            return null;
        }
    }
}

File: src/FSharpVSPowerTools.Pure/Commands/ResolveUnopenedNamespaceSmartTaggerProvider.fs
﻿module FSharpVSPowerTools.ResolveUnopenedNamespaceSmartTaggerProvider

open System
open System.ComponentModel.Composition
open Microsoft.VisualStudio.Text
open Microsoft.VisualStudio.Text.Editor
open Microsoft.VisualStudio.Text.Operations
open Microsoft.VisualStudio.Text.Tagging
open Microsoft.VisualStudio.Utilities
open Microsoft.VisualStudio.Shell
open FSharpVSPowerTools.Refactoring
open FSharpVSPowerTools.ProjectSystem
open System.Diagnostics
open Microsoft.VisualStudio.Shell.Interop

[<Export (typeof<IViewTaggerProvider>)>]
[<ContentType "F#">]
[<TagType (typeof<ResolveUnopenedNamespaceSmartTag>)>]
[<TextViewRole (PredefinedTextViewRoles.Editable)>]
type ResolveUnopenedNamespaceSmartTaggerProvider [<ImportingConstructor>]
   ([<Import (typeof<SVsServiceProvider>)>] 
    serviceProvider             :   IServiceProvider            ,
    textDocumentFactoryService  :   ITextDocumentFactoryService ,
    undoHistoryRegistry         :   ITextUndoHistoryRegistry    ,
    projectFactory              :   ProjectFactory              ,
    vsLanguageService           :   VSLanguageService           ) =

    interface IViewTaggerProvider with
        member __.CreateTagger (textView, buffer) =
            maybe {
                if textView.TextBuffer <> buffer then return! None else
                let! generalOptions = Setting.tryGetGeneralOptions ()//serviceProvider
                
                let dte = Package.GetService<SDTE,EnvDTE.DTE> ()
                if dte.Version = string VisualStudioVersion.VS2015 then return! None else
                let! doc = textDocumentFactoryService.TryDocumentFromBuffer buffer
                let resolver = 
                    new UnopenedNamespaceResolver (doc, textView, undoHistoryRegistry.RegisterHistory buffer, 
                        vsLanguageService, projectFactory)
                return
                    new ResolveUnopenedNamespaceSmartTagger (buffer, resolver) :> obj :?> _
            } |> Option.getOrElse null
            
